code/include/swoc/Vectray.h

```diff
@@ -0,0 +1,370 @@
+        this->template emplace_back(std::move(item));
+      }
+    }
+  }
+}
+
+template <typename T, size_t N, class A> Vectray<T, N, A>::FixedStore::~FixedStore() {
+  for ( auto & item : this->span() ) {
+    std::destroy_at(std::addressof(item));
+  }
+}
+
+template<typename T, size_t N, class A>
+MemSpan<T> Vectray<T, N, A>::FixedStore::span() {
+  return MemSpan(_raw).template rebind<T>();
+}
+
+template<typename T, size_t N, typename A>
+T& Vectray<T,N,A>::operator[](size_type idx) {
+  return this->items()[idx];
+}
+
+template<typename T, size_t N, typename A>
+T const& Vectray<T,N,A>::operator[](size_type idx) const {
+  return this->items[idx];
+}
+
+template<typename T, size_t N, typename A>
+auto Vectray<T,N,A>::push_back(const T& t) -> self_type& {
+  std::visit(swoc::meta::vary{
+      [&](FixedStore& fs) -> void {
+        if (fs._count < N) {
+          new(reinterpret_cast<T*>(fs._raw.data()) + fs._count++) T(t); // A::traits ?
+        } else {
+          this->transfer();
+          std::get<DYNAMIC>(_store).push_back(t);
+        }
+      }
+      , [&](DynamicStore& ds) -> void {
+        ds.push_back(t);
+      }
+  }, _store);
+  return *this;
+}
+
+template<typename T, size_t N, typename A>
+auto Vectray<T,N,A>::push_back(T && t) -> self_type& {
+  std::visit(swoc::meta::vary{
+               [&](FixedStore& fs) -> void {
+                 if (fs._count < N) {
+                   new(reinterpret_cast<T*>(fs._raw.data()) + fs._count++) T(std::move(t)); // A::traits ?
+                 } else {
+                   this->transfer();
+                   std::get<DYNAMIC>(_store).push_back(t);
+                 }
+               }
+               , [&](DynamicStore& ds) -> void {
+                 ds.push_back(std::move(t));
+               }
+             }, _store);
+  return *this;
+}
+
+template<typename T, size_t N, class A>
+template<typename... Args>
+auto Vectray<T, N, A>::emplace_back(Args && ... args) -> self_type& {
+  if (_store.index() == FIXED) {
+    auto& fs{std::get<FIXED>(_store)};
+    if (fs._count < N) {
+      new(reinterpret_cast<T*>(fs._raw.data()) + fs._count++) T(std::forward<Args>(args)...); // A::traits ?
+      return *this;
+    }
+    this->transfer(); // transfer to dynamic and fall through to add item.
+  }
+  std::get<DYNAMIC>(_store).emplace_back(std::forward<Args>(args)...);
+  return *this;
+}
+
+template<typename T, size_t N, class A>
+auto Vectray<T, N, A>::pop_back() -> self_type & {
+  std::visit(swoc::meta::vary{
+               [&](FixedStore& fs) -> void { std::destroy_at(fs.span()[--fs._count]); }
+             , [&](DynamicStore& ds) -> void { ds.pop_back(); }
+             }, _store);
+  return *this;
+}
+
+template<typename T, size_t N, typename A>
+auto Vectray<T,N,A>::size() const -> size_type {
+  return std::visit(swoc::meta::vary{
+      [](FixedStore const& fs) { return fs._count; }
+      , [](DynamicStore const& ds) { return ds.size(); }
+  }, _store);
+}
```

Comment:
This is a question:

With std::array, size() is a constexpr function which always returns `N`. If we're managed by FixedStore, what does this return? `N` or the number of items after pushes and pops, starting form 0?

I'm not sure there's a wrong answer to this, just wondering how it behaves.

Reply:
The number of valid elements. For `std::array` that's always the size of the array, but not for a vector or this class.

Reply:
Yes, that's good. I prefer it acting the way you describe.

You might want to add a sentence explicitly saying that in the document. Since this is a class that is a combination of std::array and std::vector behaviors, it wasn’t obvious to me how size() would behave in the N or smaller case.